programs/ssl/ssl_client2.c

```diff
@@ -1517,16 +1517,19 @@ int main(int argc, char *argv[])
                     *p++ = '\0';
                 }
 
-                if ((curve_cur = mbedtls_ecp_curve_info_from_name(q)) != NULL) {
-                    group_list[i++] = curve_cur->tls_id;
+                curve_tls_id = mbedtls_ssl_get_curve_tls_id_from_name(q);
```

Comment:
This function is only defined when `DEBUG_C` is, but the rest of `ssl_client2.c` can work without `DEBUG_C`. I'm not sure how to resolve this. Perhaps guard this part with `#if defined(DEBUG_C)` and
```
#else /* MBEDTLS_DEBUG_C */
mbedtls_printf("The 'curves' option is only supported with MBEDTLS_DEBUG_C\n");
ret = -1
goto exit;
#endif /* MBEDTLS_DEBUG_C */
```

Reply:
Otherwise the only alternative I can see is to duplicate the list of curve names in the program (or in `ssl_test_lib` if `ssl_server2.c` ever needs it). Actually I just checked, and it looks like it does.

Reply:
True, TBH I expected that there could be some failure due to the fact that this function is defined only when `MBEDTLS_DEBUG_C` is. The "problem" is that curves' names in `tls_id_match_table` in `ssl_tls.c` are only defined when `MBEDTLS_DEBUG_C`. This was for historical reasons since when we did this we thought that having the possibility to print curve's name was required only when `MBEDTLS_DEBUG_C` was enabled which made sense at that time.
I don't really like to duplicate the same structures in multiple places since this leads to maintainability issues usually. Why don't we simply remove `MBEDTLS_DEBUG_C` from the strings in `tls_id_match_table`?
I know that the code size is going to increase, but not so badly...

Reply:
OTOH it is also true that I don't know how many end users are really interested in knowing the names of the supported curves. Perhaps getting a list of supported TLS ID is more interesting? Then we can use the function which translates to name on each of them only when `MBEDTLS_DEBUG_C` is defined.

Reply:
As a general rule, given one of our primary purposes is to support constrained embedded devices, we really shouldn't be wasting precious flash/ROM space with human-readable strings. In the past we've included too many such strings in the library (`md.h`, `ecp.h`, etc) but in the future we shouldn't. The only exception being `DEBUG_C` as it's not supposed to be enabled in production, and its very nature means including more strings. Even then, I'm not convinced we need a human-readable name for each curve, hash, etc. - we could just include the numeric TLS ID in the output. People who read debug output are supposed to be able to translate that to a curve name by grepping the source or so.

So, I'm almost inclined to entirely remove curve name from `tls_id_match_table` (even when `DEBUG_C` is enabled, and then in `ssl_test_lib.c` have a table that just matches TLS ID to human-readable name - that way there's no duplication and strings are cleanly out of the library.

So, basically, the two functions added here should instead live in `ssl_test_lib.c` and to avoid duplication we can remove strings from `tls_id_match_table` - or just accept the duplication, as it's not withing the library, and the list of available curves doesn't change that often.